File: 120/main.py
from typing import List


class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        """
        approach
        bottom up dp, because we end up with 1 value at the top
        for each level above the bottom, set value == min(bottom left, bottom right) + current
        return triangle[0][0]

        edge cases
        0 levels -> not possible
        1 level -> return only value
        """
        for i in range(len(triangle) - 2, -1, -1):
            for j in range(len(triangle[i])):
                triangle[i][j] += min(triangle[i + 1][j], triangle[i + 1][j + 1])
        return triangle[0][0]


"""
1

1
2 3
1 4 8

4
3 7
1 4 8

"""

File: 120/main.ts
function minimumTotal(triangle: number[][]): number {
  for (let i = triangle.length - 2; i >= 0; i--) {
    let row = triangle[i];
    for (let j = 0; j < row.length; j++) {
      triangle[i][j] =
        triangle[i][j] + Math.min(triangle[i + 1][j], triangle[i + 1][j + 1]);
    }
  }
  return triangle[0][0];
}

// DP from the lowest level of the triangle
// for each level above the lowest, set value = value + min(triangle[i + 1][j], triangle[i + 1][j + 1])
// return triangle [0][0]
